modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,359 @@
+// TODO currently the cached trees are not persisted and we build them at
+// startup. For petabytes of data this might take a long time.
+
+package proto
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/errors"
+)
+
+// merkleRootsCacheHeight is the height of the subTrees in cachedSubTrees. A
+// height of 7 means that 128 sector roots are covered by a single cached
+// subTree.
+const merkleRootsCacheHeight = 7
+
+// merkleRootsPerCache is the number of merkle roots in a cached subTree of
+// merkleRootsCacheHeight height.
+const merkleRootsPerCache = 1 << merkleRootsCacheHeight
+
+type (
+	// merkleRoots is a helper struct that makes it easier to add/insert/remove
+	// merkleRoots within a SafeContract.
+	merkleRoots struct {
+		// cachedSubTrees are cached trees that can be used to more efficiently
+		// compute the merkle root of a contract.
+		cachedSubTrees []*cachedSubTree
+		// uncachedRoots contains the sector roots that are not part of a
+		// cached subTree. The uncachedRoots slice should never get longer than
+		// 2^merkleRootsCacheHeight since that would simply result in a new
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
+		// numMerkleRoots is the number of merkle roots in file.
+		numMerkleRoots int
+	}
+
+	// cachedSubTree is a cached subTree of a merkle tree. A height of 0 means
+	// that the sum is the hash of a leaf. A subTree of height 1 means sum is
+	// the root of 2 leaves. A subTree of height 2 contains the root of 4
+	// leaves and so on.
+	cachedSubTree struct {
+		height int         // height of the subTree
+		sum    crypto.Hash // root of the subTree
+	}
+)
+
+// loadExistingMerkleRoots reads creates a merkleRoots object from existing
+// merkle roots.
+func loadExistingMerkleRoots(file *os.File) (*merkleRoots, error) {
+	mr := &merkleRoots{
+		file: file,
+	}
+	// Get the number of roots stored in the file.
+	var err error
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return nil, err
+	}
+	// Seek to the first root's offset.
+	if _, err = file.Seek(fileOffsetFromRootIndex(0), io.SeekStart); err != nil {
```

Comment:
I'm thinking we want this function to be a method on the contract, instead of a global constant. I suspect that if we ever change the offset in the future, having it a method of the safecontract will clean things up when we transition from one thing to the next.

Reply:
maybe it's not such a big deal